### PR TITLE
parser: convert identifier to utf8mb4 charset

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close  https://github.com/pingcap/tidb/issues/31113

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

File: parser/charset/encoding_base.go [hunk 1]
```diff
@@ -128,3 +128,16 @@ func Slice(s string) (b []byte) {
 	pBytes.Cap = pString.Len
 	return
 }
+
+// String converts slice to string without copy.
+// Use it at your own risk.
+func String(b []byte) (s string) {
```
Comment (tangenta) [inline parser/charset/encoding_base.go:134]:
```suggestion
func HackString(b []byte) (s string) {
```

Could you prepend "Hack" to `String` and `Slice` to make it more obvious?

File: parser/lexer.go [hunk 2]
```diff
@@ -147,6 +147,18 @@ func (s *Scanner) AppendWarn(err error) {
 	s.warns = append(s.warns, err)
 }
 
+// convert2System convert lit from client encoding to system encoding which is utf8mb4.
+func (s *Scanner) convert2System(tok int, lit string) (int, string) {
+	utf8Lit, err := s.client.Transform(nil, charset.Slice(lit), charset.OpDecodeReplace)
+	if err != nil {
+		s.AppendError(err)
+		s.lastErrorAsWarn()
```
Comment (Defined2014) [inline parser/lexer.go:155]: How about `s.AppendWarn(err)`?

**Review by tangenta — APPROVED** [review]

File: parser/charset/encoding_base.go (hunk 1 above)
Comment (Defined2014) [reply parser/charset/encoding_base.go:134]: Maybe we can move String() and Slice() to other file(like `util`) later.
And delete `hack.Slice` in `tidb/util/hack/hack.go`

**Review by Defined2014 — APPROVED** [review]

**Pull request merged**